### ISSUE TITLE
Additional headers/data for Ajax calls for tiles

### ISSUE THREAD

**Issue by ChiSamurai** [openseadragon/openseadragon#1032]
Hi all,

we have most of our images (using IIIF) distributed on serveral servers with restricted access. To use OSD for display it is needed to do a (remote) login which throws back an auth token and finally pass this token as a HTTP-header in the ajax calls (for fetching info.json as well as the tiles). 

According to @avandecreme on Gitter at the moment there is no option for this, but a fork which took on this: https://github.com/seanich/openseadragon

Cheers,
Matthias

**Comment by avandecreme** [comment]
@seanich Any feedback on your fork? Is it working well? Is there any reason we shouldn't always use Ajax? One reason I can think of is to support `file://` URLs, so we should probably let the user choose.

Would you be ok to make a PR to support this?

**Comment by seanich** [comment]
My fork is working quite well. I haven't noticed any performance differences between loading tiles via Ajax vs. loading via the Image element. I haven't done any objective testing to support this, though.

Attaching headers to tile requests is definitely a useful feature both for @ChiSamurai authentication use case as well as retrieving partial content via the `Content-Range` header from something like Google Cloud Storage.

I think that loading tiles via Ajax should be an opt-in configuration option (such as `loadTilesWithAjax`). In addition to @avandecreme's point about `file://` URLs, the use of [`responseType = 'arraybuffer'`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#Browser_compatibility) and [createObjectURL](http://caniuse.com/#feat=bloburls) would limit support to IE 10+.

I'd be happy to work on a PR to support this feature behind a configuration option since there's some interest in it :).

Here's the current plan for implementation in case anyone has comments or suggestions:

* Add the `loadTilesWithAjax` configuration option which defaults to `false`
* Add the [`getTileHeaders`](https://github.com/seanich/openseadragon/blob/master/src/tilesource.js#L562) method to the TileSource interface.

**Comment by avandecreme** [comment]
Thanks for your feedback!
Good thought about content-range. This will be needed by @eriksjolund in #944.

Your implementation suggestion with the opt in option sounds good to me.
Regarding the getTileHeaders in TileSource, my first thought is that it would be better as an option to viewport.addTiledImage. This avoid to reimplement a new TileSource for each different header.
What do you think?

**Comment by seanich** [comment]
My thought about adding the `getTileHeaders` function to the tile source is that it would have the same signature as [`getTileUrl`](http://openseadragon.github.io/docs/OpenSeadragon.TileSource.html#getTileUrl). This would allow you to look up a specific tile's byte range in whatever data structure you might be storing that in. 

For example, if you had a two dimensional array storing the byte offsets of each tile you might implement `getTileHeaders` like so:

```javascript
getTileHeaders(level, x, y) {
  const tilesX = Math.ceil((this.getLevelScale(level) * this.dimensions.x) / this.getTileWidth());
  const tileIndex = y * tilesX + x;
  const bytesStart = tileByteOffsets[level][tileIndex];
  const bytesEnd = tileByteOffsets[level][tileIndex];
  const bytesTotal = tileByteOffsets[level][tileByteOffsets[level].length - 1];

  return {
    'Content-Range': `bytes ${bytesStart}-${bytesEnd}/${bytesTotal}`
  };
}
```

Of course that implementation makes a lot of unnecessary calculations that should be cached somewhere, but hopefully it illustrates the point.

However, this method doesn't really make sense for headers that don't change based on which tile is being requested (e.g. auth headers). For that purpose it might make sense to add a configuration option (such as `tileRequestHeaders`) to either the `Viewer` constructor or the `TiledImage` constructor and merge the headers set there with those returned from `getTileHeaders` (if any).

Something like

```javascript
const myViewer = new OpenSeaDragon.Viewer({
  // ...
  loadTilesWithAjax: true,
  tileRequestHeaders: {
    Authentication: 'Bearer <AUTH_TOKEN>'
  },
  // ...
});
```

or 

```javascript
myViewer.addTiledImage({
  //...
  tileRequestHeaders: {
    Authentication: 'Bearer <AUTH_TOKEN>'
  }
  //...
})
```

Naming things isn't my strong suit, so any suggestions on the name of options is certainly welcome.

**Comment by ChiSamurai** [comment]
Hm that all sounds really nice to me. One thought: would it be also possible to have a callback function if the ajax call was unauthorized? Or better: a callback with the complete request answer so one could i.e. react to a 401 by presenting a login formular or similar?

**Comment by seanich** [comment]
Rather than providing a callback option we can extend the existing [`tile-load-failed`](http://openseadragon.github.io/docs/OpenSeadragon.Viewer.html#.event:tile-load-failed) event to include extra information (e.g. status code, response body, etc.) when available.

**Comment by avandecreme** [comment]
All that sounds good to me. Thanks for the clarification.
Maybe @iangilman will have some other thoughts when he is back.

**Comment by seanich** [comment]
Great! I'll work on a PR this weekend so we can look over the code changes that need to be made and we can discuss things from there.

**Comment by seanich** [comment]
Just wanted to provide an update to anyone that's interested in this issue. I haven't had a whole lot of time to work on the PR, but I have started it and I'm about half way done. I want to make sure I get the documentation and tests updated appropriately. I'll definitely have something ready by Friday. :)

**Comment by iangilman** [comment]
I'm back, and all of this sounds great to me!

**Comment by iangilman** [comment]
@seanich Ping! :)

**Comment by rahuldeepattri** [comment]
Hi @seanich, would it be possible that you could provide a fix for #1748.
